packages/synapse-react-client/src/components/useImmutableTableQuery.ts

```diff
@@ -86,16 +89,34 @@ export default function useImmutableTableQuery(
   options: UseImmutableTableQueryOptions,
 ): ImmutableTableQueryResult {
   const {
-    initQueryRequest,
+    initQueryRequest: initQueryRequestFromProps,
     componentIndex = 0,
     shouldDeepLink = false,
     onQueryChange,
     requireConfirmationOnChange = false,
   } = options
 
-  const [lastQueryRequest, setLastQueryRequest] =
+  const initQueryRequest = useMemo(() => {
+    const request = cloneDeep(initQueryRequestFromProps)
+    request.query = removeEmptyQueryParams(request.query)
+    return request
+  }, [initQueryRequestFromProps])
+
+  const [lastQueryRequest, setLastQueryRequestInternal] =
     useState<QueryBundleRequest>(initQueryRequest)
 
+  function setLastQueryRequest(
+    newState: React.SetStateAction<QueryBundleRequest>,
+  ) {
+    setLastQueryRequestInternal(newState)
+    const request =
+      typeof newState === 'function' ? newState(lastQueryRequest) : newState
+
+    if (onQueryChange) {
+      onQueryChange(JSON.stringify(request.query))
+    }
```

Comment:
`onQueryChange` is now called when `setLastQueryRequest` is called, which is the change event.

Before, `onQueryChange` was called in a `useEffect` block, but we didn't want it to run on every change. For example, it shouldn't run on mount.

This will be replaced by new, similar code once `useCommittedState` from #382 is integrated.